Added ‘Notes’ to ‘Job’ so users can record some notes about each job …

steven-hill opened the pull request:
…application.

Adding a new job application to the tracker:
	- tapping on ‘+’ in HomeViewController will present a sheet (AddNewJobViewController).
	- users can enter text for the company, position, location, choose a status from the popup button, and add notes.

Editing a previously saved job application:
	- users can edit a job application by tapping on its cell.
	- the job application details will be presented in a sheet (JobDetailViewController).
	- users can make changes and tap ‘Done’ to save them or ‘Cancel’ to discard them.

Added a status label to both sheets to indicate that the status can be chosen from the popup button beside it.

The dashboard now shows dynamic data and updates when new applications are added or when existing ones are edited.

Implemented CoreData for persistence. If there is a CoreData error, an alert is shown to the user.

Now that the data comes from CoreData, the ‘Job’ file was deleted. This file was only used for dummy data while building the UI.

Added search functionality by company name:
    - the app gets the search query from the search bar.
    - it performs a fetch request using the query string.
    - it assigns the results to jobsArray and displays the results in the collection view.
    - while the search results are shown, a button opposite the ‘Applications’ label is visible and can be tapped to allow users to go back to the main list.
    - this button is only visible to the user when search results are displayed in the collection view.
    - if the search query returns no results, an alert notifies the user.

OlivAlex commented on Job Application Tracker/JobApplicationTracker/JobApplicationTracker/AppDelegate.swift:
```diff
@@ -9,28 +9,17 @@ import UIKit
         return UISceneConfiguration(name: "Default Configuration", sessionRole: connectingSceneSession.role)
     }
 
     func application(_ application: UIApplication, didDiscardSceneSessions sceneSessions: Set<UISceneSession>) {
```
Can we remove all these empty methods altogether? 🤔

OlivAlex commented on Job Application Tracker/JobApplicationTracker/JobApplicationTracker/CoreData/CoreDataManager.swift:
```diff
@@ -0,0 +1,46 @@
+        return CoreDataManager.manager.persistentContainer.viewContext
+    }
+    
+    init() {
```
Do we need the empty init?

OlivAlex commented on Job Application Tracker/JobApplicationTracker/JobApplicationTracker/AppDelegate.swift:
```diff
@@ -9,28 +9,17 @@ import UIKit
 
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
-
```
❤️ Love the cleanup!

OlivAlex commented on Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/AddNewJobViewController.swift:
```diff
@@ -0,0 +1,127 @@
+        self.dismiss(animated: true, completion: nil)
+    }
+    
+    func layoutViews() {
```
I see we're setting a lot of top, leading, trailing, height constraints. Could we maybe add the UI element to an array and loop over to set those constraints?

OlivAlex commented on Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/HomeViewController.swift:
```diff
@@ -0,0 +1,258 @@
+            return
+        }
+        
+        let queryText = searchBar.text!
```
Could we add this to the guard above so we don't rely on forcefully unwrapping?

OlivAlex commented on Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/HomeViewController.swift:
```diff
@@ -0,0 +1,258 @@
+            }
+        } catch {
+            print(error)
+            DispatchQueue.main.async {
```
I see we're presenting a lot of alerts, could we maybe have a func that takes in an enum of our possible errors and presents the right alert UI?

OlivAlex commented on Job Application Tracker/JobApplicationTracker/JobApplicationTracker/ViewControllers/JobDetailViewController.swift:
```diff
@@ -0,0 +1,162 @@
+import UIKit
+import CoreData
+
+protocol JobDetailViewControllerDelegate: AnyObject {
```
Do we need the AnyObject here and in the other protocols?

OlivAlex commented:
LGTM! Lets just look at these comments and see if we can clean it up a bit! :)

steven-hill commented:
Changes made post-PR review. 

(A new commit has been pushed to the same branch.)

Removed empty methods from AppDelegate and SceneDelegate.

Removed line spaces between the IBOutlets.

Removed the isTextEntered variable from HomeViewController and rewrote the guard statement in makeSearchQuery() to remove the forcibly unwrapped optionals.

Added the missing ‘private’ keyword to the configure() method in the PopUpButton file.

Removed ‘AnyObject’ from AddNewJobViewControllerDelegate and JobDetailViewControllerDelegate.

Removed the empty init() method in the CoreDataManager file.

Re-indented code in the three view controllers.

Updated the constraints on the storyboard for the ‘Dashboard’ and ‘Applications’ labels, and ‘ReturnToAllButton’. The previous constraints cut off some of the text in the labels on iPhone SE. Now the entire text of both labels is visible on iPhone SE.

OlivAlex reviewed (approved):
This looks great. Thanks for this!
